File: src/typegraph/rust/typegraph/src/providers/temporal.rs
// Copyright Metatype OÜ, licensed under the Elastic License 2.0.
// SPDX-License-Identifier: Elastic-2.0

use crate::{
    t::{TypeBuilder, TypeDef},
    wasm::{
        self,
        core::RuntimeId,
        runtimes::{TemporalOperationData, TemporalOperationType},
    },
    Result,
};

pub use crate::wasm::runtimes::TemporalRuntimeData;

#[derive(Debug)]
pub struct TemporalRuntime {
    id: RuntimeId,
    #[allow(unused)]
    options: TemporalRuntimeData,
}

impl TemporalRuntime {
    pub fn new(options: TemporalRuntimeData) -> Result<Self> {
        let id = wasm::with_runtimes(|r, s| r.call_register_temporal_runtime(s, &options))?;

        Ok(Self { id, options })
    }

    pub fn start_workflow<T: TypeBuilder>(&self, workflow_type: &str, arg: T) -> Result<TypeDef> {
        self.operation(
            TemporalOperationType::StartWorkflow,
            Some(workflow_type),
            Some(arg.into_id()?),
            None,
        )
    }

    pub fn signal_workflow<T: TypeBuilder>(&self, signal_name: &str, arg: T) -> Result<TypeDef> {
        self.operation(
            TemporalOperationType::SignalWorkflow,
            Some(signal_name),
            Some(arg.into_id()?),
            None,
        )
    }

    pub fn query_workflow<I, O>(&self, signal_name: &str, arg: I, out: O) -> Result<TypeDef>
    where
        I: TypeBuilder,
        O: TypeBuilder,
    {
        self.operation(
            TemporalOperationType::QueryWorkflow,
            Some(signal_name),
            Some(arg.into_id()?),
            Some(out.into_id()?),
        )
    }

    pub fn describe_workflow(&self) -> Result<TypeDef> {
        self.operation(TemporalOperationType::DescribeWorkflow, None, None, None)
    }

    fn operation(
        &self,
        operation: TemporalOperationType,
        mat_arg: Option<&str>,
        func_arg: Option<u32>,
        func_out: Option<u32>,
    ) -> Result<TypeDef> {
        let data = TemporalOperationData {
            mat_arg: mat_arg.map(|arg| arg.to_string()),
            func_arg,
            func_out,
            operation,
        };

        let params =
            wasm::with_runtimes(|r, s| r.call_generate_temporal_operation(s, self.id, &data))?;

        params.build()
    }
}

File: src/typegraph/python/typegraph/providers/temporal.py
# Copyright Metatype OÜ, licensed under the Mozilla Public License Version 2.0.
# SPDX-License-Identifier: MPL-2.0

from typing import Union, Optional

from typegraph.runtimes.base import Runtime

from typegraph.gen.runtimes import (
    TemporalOperationData,
    TemporalRuntimeData,
    TemporalOperationType,
)
from typegraph.sdk import runtimes

from typegraph import t


class TemporalRuntime(Runtime):
    name: str
    host_secret: str
    namespace_secret: Optional[str]

    def __init__(
        self, name: str, host_secret: str, *, namespace_secret: Optional[str] = None
    ):
        data = TemporalRuntimeData(
            name=name, host_secret=host_secret, namespace_secret=namespace_secret
        )
        runtime_id = runtimes.register_temporal_runtime(data)
        super().__init__(runtime_id)
        self.name = name
        self.host_secret = host_secret

    def _generic_temporal_func(
        self,
        operation: TemporalOperationType,
        mat_arg: Union[None, str] = None,
        func_arg: Union[None, t.typedef] = None,
        func_out: Union[None, t.typedef] = None,
    ):
        data = TemporalOperationData(
            mat_arg=mat_arg,
            func_arg=None if func_arg is None else func_arg._id,
            func_out=None if func_out is None else func_out._id,
            operation=operation,
        )
        func_data = runtimes.generate_temporal_operation(self.id, data)
        return t.func.from_type_func(func_data)

    def start_workflow(self, workflow_type: str, arg: t.typedef):
        return self._generic_temporal_func("start_workflow", workflow_type, arg)

    def signal_workflow(self, signal_name: str, arg: t.typedef):
        return self._generic_temporal_func(
            "signal_workflow",
            signal_name,
            arg,
        )

    def query_workflow(self, query_type: str, arg: t.typedef, out: t.typedef):
        return self._generic_temporal_func("query_workflow", query_type, arg, out)

    def describe_workflow(self):
        return self._generic_temporal_func("describe_workflow")
